File: backend/app/dama/engine.py
"""Ethiopian-style Dama rules (mirrors frontend/src/games/dama/engine.ts)."""

from __future__ import annotations

from copy import deepcopy
from typing import Literal

Side = Literal["red", "black"]
PieceKind = Literal["man", "king"]

BOARD_SIZE = 8
SQUARE_COUNT = BOARD_SIZE * BOARD_SIZE

Piece = dict  # {side, kind}
Board = list  # length 64, Piece | None
Move = dict  # {from, to, captures, path, promote}


def idx(row: int, col: int) -> int:
    return row * BOARD_SIZE + col


def row_of(i: int) -> int:
    return i // BOARD_SIZE


def col_of(i: int) -> int:
    return i % BOARD_SIZE


def opposite(side: Side) -> Side:
    return "black" if side == "red" else "red"


def is_playable(i: int) -> bool:
    return (row_of(i) + col_of(i)) % 2 == 1


def create_initial_board() -> Board:
    board: Board = [None] * SQUARE_COUNT
    for i in range(SQUARE_COUNT):
        if not is_playable(i):
            continue
        r = row_of(i)
        if r < 3:
            board[i] = {"side": "black", "kind": "man"}
        elif r > 4:
            board[i] = {"side": "red", "kind": "man"}
    return board


def _in_bounds(row: int, col: int) -> bool:
    return 0 <= row < BOARD_SIZE and 0 <= col < BOARD_SIZE


def _clone(board: Board) -> Board:
    return deepcopy(board)


def _man_move_dirs(side: Side) -> list[tuple[int, int]]:
    return [(-1, -1), (-1, 1)] if side == "red" else [(1, -1), (1, 1)]


MAN_CAPTURE_DIRS = [(-1, -1), (-1, 1), (1, -1), (1, 1)]
KING_DIRS = [(-1, -1), (-1, 1), (1, -1), (1, 1)]


def _should_promote(side: Side, to: int) -> bool:
    r = row_of(to)
    return r == 0 if side == "red" else r == BOARD_SIZE - 1


def _collect_captures(
    board: Board,
    from_sq: int,
    side: Side,
    kind: PieceKind,
    origin: int,
    landings: list[int],
    captured_ordered: list[int],
    captured_set: set[int],
    results: list[Move],
) -> None:
    dirs = KING_DIRS if kind == "king" else MAN_CAPTURE_DIRS
    fr, fc = row_of(from_sq), col_of(from_sq)

    for dr, dc in dirs:
        if kind == "man":
            er, ec = fr + dr, fc + dc
            lr, lc = fr + 2 * dr, fc + 2 * dc
            if not _in_bounds(er, ec) or not _in_bounds(lr, lc):
                continue

            enemy_idx = idx(er, ec)
            land_idx = idx(lr, lc)
            enemy = board[enemy_idx]
            if not enemy or enemy["side"] == side:
                continue
            if enemy_idx in captured_set or board[land_idx] is not None:
                continue

            next_board = _clone(board)
            next_board[from_sq] = None
            next_board[enemy_idx] = None

            next_captured_ordered = [*captured_ordered, enemy_idx]
            next_captured_set = set(captured_set)
            next_captured_set.add(enemy_idx)
            next_landings = [*landings, land_idx]
            promotes = _should_promote(side, land_idx)

            if promotes:
                next_board[land_idx] = {"side": side, "kind": "king"}
                results.append(
                    {
                        "from": origin,
                        "to": land_idx,
                        "captures": next_captured_ordered,
                        "path": next_landings,
                        "promote": True,
                    }
                )
                continue

            next_board[land_idx] = {"side": side, "kind": "man"}
            before = len(results)
            _collect_captures(
                next_board,
                land_idx,
                side,
                "man",
                origin,
                next_landings,
                next_captured_ordered,
                next_captured_set,
                results,
            )
            if len(results) == before:
                results.append(
                    {
                        "from": origin,
                        "to": land_idx,
                        "captures": next_captured_ordered,
                        "path": next_landings,
                        "promote": False,
                    }
                )
        else:
            er, ec = fr + dr, fc + dc
            while _in_bounds(er, ec) and board[idx(er, ec)] is None:
                er += dr
                ec += dc
            if not _in_bounds(er, ec):
                continue

            enemy_idx = idx(er, ec)
            enemy = board[enemy_idx]
            if not enemy or enemy["side"] == side or enemy_idx in captured_set:
                continue

            lr, lc = er + dr, ec + dc
            while _in_bounds(lr, lc) and board[idx(lr, lc)] is None:
                land_idx = idx(lr, lc)
                next_board = _clone(board)
                next_board[from_sq] = None
                next_board[enemy_idx] = None
                next_board[land_idx] = {"side": side, "kind": "king"}

                next_captured_ordered = [*captured_ordered, enemy_idx]
                next_captured_set = set(captured_set)
                next_captured_set.add(enemy_idx)
                next_landings = [*landings, land_idx]

                before = len(results)
                _collect_captures(
                    next_board,
                    land_idx,
                    side,
                    "king",
                    origin,
                    next_landings,
                    next_captured_ordered,
                    next_captured_set,
                    results,
                )
                if len(results) == before:
                    results.append(
                        {
                            "from": origin,
                            "to": land_idx,
                            "captures": next_captured_ordered,
                            "path": next_landings,
                            "promote": False,
                        }
                    )
                lr += dr
                lc += dc


def _quiet_moves(board: Board, from_sq: int, side: Side, kind: PieceKind) -> list[Move]:
    moves: list[Move] = []
    fr, fc = row_of(from_sq), col_of(from_sq)
    dirs = KING_DIRS if kind == "king" else _man_move_dirs(side)

    for dr, dc in dirs:
        if kind == "man":
            nr, nc = fr + dr, fc + dc
            if not _in_bounds(nr, nc):
                continue
            to = idx(nr, nc)
            if board[to] is not None:
                continue
            moves.append(
                {
                    "from": from_sq,
                    "to": to,
                    "captures": [],
                    "path": [to],
                    "promote": _should_promote(side, to),
                }
            )
        else:
            nr, nc = fr + dr, fc + dc
            while _in_bounds(nr, nc) and board[idx(nr, nc)] is None:
                to = idx(nr, nc)
                moves.append(
                    {
                        "from": from_sq,
                        "to": to,
                        "captures": [],
                        "path": [to],
                        "promote": False,
                    }
                )
                nr += dr
                nc += dc
    return moves


def legal_moves(board: Board, side: Side) -> list[Move]:
    captures: list[Move] = []
    quiet: list[Move] = []

    for i in range(SQUARE_COUNT):
        piece = board[i]
        if not piece or piece["side"] != side:
            continue

        piece_captures: list[Move] = []
        _collect_captures(
            board, i, side, piece["kind"], i, [], [], set(), piece_captures
        )
        captures.extend(piece_captures)

        if not piece_captures:
            quiet.extend(_quiet_moves(board, i, side, piece["kind"]))

    if captures:
        max_len = max(len(m["captures"]) for m in captures)
        return [m for m in captures if len(m["captures"]) == max_len]
    return quiet


def apply_move(board: Board, move: Move) -> Board:
    next_board = _clone(board)
    piece = next_board[move["from"]]
    if not piece:
        return next_board

    next_board[move["from"]] = None
    for c in move["captures"]:
        next_board[c] = None

    kind = piece["kind"]
    if move.get("promote") or _should_promote(piece["side"], move["to"]):
        kind = "king"
    next_board[move["to"]] = {"side": piece["side"], "kind": kind}
    return next_board


def evaluate_outcome(board: Board, side_to_move: Side) -> Side | Literal["draw"] | None:
    moves = legal_moves(board, side_to_move)
    if moves:
        return None

    has_red = any(p and p["side"] == "red" for p in board)
    has_black = any(p and p["side"] == "black" for p in board)
    if not has_red and not has_black:
        return "draw"
    if not has_red:
        return "black"
    if not has_black:
        return "red"
    return opposite(side_to_move)


def find_legal_move(board: Board, side: Side, from_sq: int, to_sq: int) -> Move | None:
    for move in legal_moves(board, side):
        if move["from"] == from_sq and move["to"] == to_sq:
            return move
    return None

File: frontend/src/games/dama/engine.ts
/** Ethiopian-style Dama (checkers) on an 8×8 board with flying kings. */

export type Side = "red" | "black";
export type PieceKind = "man" | "king";

export interface Piece {
  side: Side;
  kind: PieceKind;
}

/** Board is row-major; null = empty. Dark playable squares only matter for UI. */
export type Board = (Piece | null)[];

export interface Move {
  from: number;
  to: number;
  /** Captured square indices in jump order (empty for quiet moves). */
  captures: number[];
  /** Landing squares visited after `from` (for multi-jump animation). */
  path: number[];
  promote: boolean;
}

export type GameMode = "ai" | "local" | "online";

export const BOARD_SIZE = 8;
export const SQUARE_COUNT = BOARD_SIZE * BOARD_SIZE;

export function idx(row: number, col: number): number {
  return row * BOARD_SIZE + col;
}

export function rowOf(i: number): number {
  return Math.floor(i / BOARD_SIZE);
}

export function colOf(i: number): number {
  return i % BOARD_SIZE;
}

export function opposite(side: Side): Side {
  return side === "red" ? "black" : "red";
}

/** Dark squares are playable (row+col odd). */
export function isPlayable(i: number): boolean {
  const r = rowOf(i);
  const c = colOf(i);
  return (r + c) % 2 === 1;
}

export function createInitialBoard(): Board {
  const board: Board = Array(SQUARE_COUNT).fill(null);

  for (let i = 0; i < SQUARE_COUNT; i++) {
    if (!isPlayable(i)) continue;
    const r = rowOf(i);
    if (r < 3) board[i] = { side: "black", kind: "man" };
    else if (r > 4) board[i] = { side: "red", kind: "man" };
  }

  return board;
}

function inBounds(row: number, col: number): boolean {
  return row >= 0 && row < BOARD_SIZE && col >= 0 && col < BOARD_SIZE;
}

function cloneBoard(board: Board): Board {
  return board.map((p) => (p ? { ...p } : null));
}

/** Quiet moves for men stay forward only. */
function manMoveDirs(side: Side): Array<[number, number]> {
  return side === "red"
    ? [
        [-1, -1],
        [-1, 1],
      ]
    : [
        [1, -1],
        [1, 1],
      ];
}

/** Men may capture in all four diagonal directions (back-eating allowed). */
const MAN_CAPTURE_DIRS: Array<[number, number]> = [
  [-1, -1],
  [-1, 1],
  [1, -1],
  [1, 1],
];

const KING_DIRS: Array<[number, number]> = [
  [-1, -1],
  [-1, 1],
  [1, -1],
  [1, 1],
];

function shouldPromote(side: Side, to: number): boolean {
  const r = rowOf(to);
  return side === "red" ? r === 0 : r === BOARD_SIZE - 1;
}

/**
 * Collect captures for a piece. Men jump adjacent enemies (any diagonal).
 * Kings fly any distance. If a man lands on the crowning row mid-sequence,
 * the move ends there — no immediate king capture on the same turn.
 */
function collectCaptures(
  board: Board,
  from: number,
  side: Side,
  kind: PieceKind,
  origin: number,
  landings: number[],
  capturedOrdered: number[],
  capturedSet: Set<number>,
  results: Move[],
): void {
  let extended = false;
  const dirs = kind === "king" ? KING_DIRS : MAN_CAPTURE_DIRS;
  const fr = rowOf(from);
  const fc = colOf(from);

  for (const [dr, dc] of dirs) {
    if (kind === "man") {
      const er = fr + dr;
      const ec = fc + dc;
      const lr = fr + 2 * dr;
      const lc = fc + 2 * dc;
      if (!inBounds(er, ec) || !inBounds(lr, lc)) continue;

      const enemyIdx = idx(er, ec);
      const landIdx = idx(lr, lc);
      const enemy = board[enemyIdx];
      if (!enemy || enemy.side === side) continue;
      if (capturedSet.has(enemyIdx)) continue;
      if (board[landIdx] !== null) continue;

      const nextBoard = cloneBoard(board);
      nextBoard[from] = null;
      nextBoard[enemyIdx] = null;

      const nextCapturedOrdered = [...capturedOrdered, enemyIdx];
      const nextCapturedSet = new Set(capturedSet);
      nextCapturedSet.add(enemyIdx);
      const nextLandings = [...landings, landIdx];
      const promotes = shouldPromote(side, landIdx);

      if (promotes) {
        // Become king and end the turn — no further eating this move.
        nextBoard[landIdx] = { side, kind: "king" };
        results.push({
          from: origin,
          to: landIdx,
          captures: nextCapturedOrdered,
          path: nextLandings,
          promote: true,
        });
        extended = true;
        continue;
      }

      nextBoard[landIdx] = { side, kind: "man" };
      const before = results.length;
      collectCaptures(
        nextBoard,
        landIdx,
        side,
        "man",
        origin,
        nextLandings,
        nextCapturedOrdered,
        nextCapturedSet,
        results,
      );
      if (results.length === before) {
        results.push({
          from: origin,
          to: landIdx,
          captures: nextCapturedOrdered,
          path: nextLandings,
          promote: false,
        });
      }
      extended = true;
    } else {
      let er = fr + dr;
      let ec = fc + dc;
      while (inBounds(er, ec) && board[idx(er, ec)] === null) {
        er += dr;
        ec += dc;
      }
      if (!inBounds(er, ec)) continue;

      const enemyIdx = idx(er, ec);
      const enemy = board[enemyIdx];
      if (!enemy || enemy.side === side) continue;
      if (capturedSet.has(enemyIdx)) continue;

      let lr = er + dr;
      let lc = ec + dc;
      while (inBounds(lr, lc) && board[idx(lr, lc)] === null) {
        const landIdx = idx(lr, lc);
        const nextBoard = cloneBoard(board);
        nextBoard[from] = null;
        nextBoard[enemyIdx] = null;
        nextBoard[landIdx] = { side, kind: "king" };

        const nextCapturedOrdered = [...capturedOrdered, enemyIdx];
        const nextCapturedSet = new Set(capturedSet);
        nextCapturedSet.add(enemyIdx);
        const nextLandings = [...landings, landIdx];

        const before = results.length;
        collectCaptures(
          nextBoard,
          landIdx,
          side,
          "king",
          origin,
          nextLandings,
          nextCapturedOrdered,
          nextCapturedSet,
          results,
        );
        if (results.length === before) {
          results.push({
            from: origin,
            to: landIdx,
            captures: nextCapturedOrdered,
            path: nextLandings,
            promote: false,
          });
        }
        extended = true;
        lr += dr;
        lc += dc;
      }
    }
  }

  void extended;
}

function quietMovesFor(board: Board, from: number, side: Side, kind: PieceKind): Move[] {
  const moves: Move[] = [];
  const fr = rowOf(from);
  const fc = colOf(from);
  const dirs = kind === "king" ? KING_DIRS : manMoveDirs(side);

  for (const [dr, dc] of dirs) {
    if (kind === "man") {
      const nr = fr + dr;
      const nc = fc + dc;
      if (!inBounds(nr, nc)) continue;
      const to = idx(nr, nc);
      if (board[to] !== null) continue;
      moves.push({
        from,
        to,
        captures: [],
        path: [to],
        promote: shouldPromote(side, to),
      });
    } else {
      let nr = fr + dr;
      let nc = fc + dc;
      while (inBounds(nr, nc) && board[idx(nr, nc)] === null) {
        const to = idx(nr, nc);
        moves.push({
          from,
          to,
          captures: [],
          path: [to],
          promote: false,
        });
        nr += dr;
        nc += dc;
      }
    }
  }

  return moves;
}

/** All legal moves for ``side``. Captures are mandatory when available. */
export function legalMoves(board: Board, side: Side): Move[] {
  const captures: Move[] = [];
  const quiet: Move[] = [];

  for (let i = 0; i < SQUARE_COUNT; i++) {
    const piece = board[i];
    if (!piece || piece.side !== side) continue;

    const pieceCaptures: Move[] = [];
    collectCaptures(
      board,
      i,
      side,
      piece.kind,
      i,
      [],
      [],
      new Set(),
      pieceCaptures,
    );
    captures.push(...pieceCaptures);

    if (pieceCaptures.length === 0) {
      quiet.push(...quietMovesFor(board, i, side, piece.kind));
    }
  }

  if (captures.length > 0) {
    const maxLen = Math.max(...captures.map((m) => m.captures.length));
    return captures.filter((m) => m.captures.length === maxLen);
  }

  return quiet;
}

export function applyMove(board: Board, move: Move): Board {
  const next = cloneBoard(board);
  const piece = next[move.from];
  if (!piece) return next;

  next[move.from] = null;
  for (const c of move.captures) next[c] = null;

  let kind = piece.kind;
  if (move.promote || shouldPromote(piece.side, move.to)) {
    kind = "king";
  }
  next[move.to] = { side: piece.side, kind };
  return next;
}

export type Outcome = "red" | "black" | "draw" | null;

export function evaluateOutcome(board: Board, sideToMove: Side): Outcome {
  const moves = legalMoves(board, sideToMove);
  if (moves.length > 0) return null;

  const hasRed = board.some((p) => p?.side === "red");
  const hasBlack = board.some((p) => p?.side === "black");
  if (!hasRed && !hasBlack) return "draw";
  if (!hasRed) return "black";
  if (!hasBlack) return "red";
  return opposite(sideToMove);
}

export function countPieces(board: Board, side: Side): { men: number; kings: number } {
  let men = 0;
  let kings = 0;
  for (const p of board) {
    if (!p || p.side !== side) continue;
    if (p.kind === "king") kings += 1;
    else men += 1;
  }
  return { men, kings };
}

export function formatClock(ms: number): string {
  const total = Math.max(0, Math.floor(ms / 1000));
  const m = Math.floor(total / 60);
  const s = total % 60;
  return `${m}:${s.toString().padStart(2, "0")}`;
}
